### ISSUE TITLE
Git SSH keys failing to authenticate

### ISSUE THREAD

**Issue by btrepp** [argoproj/argo-cd#9122]
Checklist:

* [ ] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Argocd is failing to authenticate with a valid private key
```
argocd@argocd-server-c9958cfc8-cwvg2:~$ argocd repo add git@git.sr.ht:/~btrepp/infrastructure --ssh-private-key-path ./id --loglevel debug
FATA[0002] rpc error: code = Unknown desc = error testing repository connectivity: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain 
```
Using ssh on the same server shows the key is good
```
argocd@argocd-server-c9958cfc8-cwvg2:~$ ssh -i id git@git.sr.ht 
PTY allocation request failed on channel 0
Hi btrepp! You've successfully authenticated, but I do not provide an interactive shell. Bye!
Connection to git.sr.ht closed.
```

**To Reproduce**

Run `argocd add repo`

**Expected behavior**

Argocd adds the repo with it's credentials

**Screenshots**

**Version**

```shell
argocd@argocd-server-c9958cfc8-cwvg2:~$ argocd version
argocd: v2.3.3+07ac038
  BuildDate: 2022-03-30T00:59:49Z
  GitCommit: 07ac038a8f97a93b401e824550f0505400a8c84e
  GitTreeState: clean
  GoVersion: go1.17.6
  Compiler: gc
  Platform: linux/arm64
argocd-server: v2.3.3+07ac038
  BuildDate: 2022-03-30T00:59:49Z
  GitCommit: 07ac038a8f97a93b401e824550f0505400a8c84e
  GitTreeState: clean
  GoVersion: go1.17.6
  Compiler: gc
  Platform: linux/arm64
  Ksonnet Version: vdev-2022-03-30T00:42:07+0000
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.0+gd141386
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by nikhleshgoenka** [comment]
Did you try cloning the respiratory with same private key?

ssh-agent
ssh-add <pvt-key>
git clone <repo-name>

**Comment by btrepp** [comment]
Clones fine
```
argocd@argocd-server-7d74dbfbc7-4vklw:~$ eval $(ssh-agent)
Agent pid 66
argocd@argocd-server-7d74dbfbc7-4vklw:~$ ssh-add id
Identity added: id (beau@beau-pc)
argocd@argocd-server-7d74dbfbc7-4vklw:~$ git clone git@git.sr.ht:/~btrepp/infrastructure
Cloning into 'infrastructure'...
remote: Enumerating objects: 1490, done.
remote: Counting objects: 100% (1048/1048), done.
remote: Compressing objects: 100% (1014/1014), done.
remote: Total 1490 (delta 610), reused 0 (delta 0), pack-reused 442
Receiving objects: 100% (1490/1490), 297.37 KiB | 378.00 KiB/s, done.
Resolving deltas: 100% (776/776), done.
```

**Comment by btrepp** [comment]
```
ssh-keygen -t ed25519 -C "argocd" -f argocd
```
from #7600 
Has allowed me to finally craft a key argocd will use, it's not the same host (e.g not github).

I think argocd isn't supporting the default flags of whatever ssh-keygen arch installs.

**Comment by protosam** [comment]
Argo CD only half supports PEM keys. The error is confusing because of history like #2890 and from that history you'll notice mention that Argo does not consistently handle the git URLs and the disparity is visible when attempting to use PEM formatted private keys, which you'll have if you're migrating from fluxcd.

The error presents as follows for the application:
```
rpc error: code = Internal desc = Failed to fetch default: `git fetch origin --tags --force` failed exit status 128: Load key "/dev/shm/2504980542": invalid format git@github.com: Permission denied (publickey). fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists.
```

Viewing the repo status, it has "Successful" for the status.

**Comment by MaiCabrera** [comment]
any one have update/workaround and or a solution on this bug?  
I've been literally suck in the same situation. @protosam  did you get any luck changing the ssh key ?

**Comment by protosam** [comment]
@MaiCabrera yes, just don't use PEM formatted keys. Use OpenSSH formatted keys. These commands will change the files in-place. Make backups beforehand just in case.
```
# to openssh format
ssh-keygen -e -p -f ~/.ssh/id_rsa-keyname -m RFC4716

# to pem format
ssh-keygen -e -p -f ~/.ssh/id_rsa-keyname -m PEM
```

I actually stopped using Argo CD after seeing these inconsistencies and have fully adopted Flux v2 at this point. The one thing Argo CD would do for me that Flux v2 doesn't is accept an Application that installs cert-manager via helm + Issuers, and eventually reconcile these things. 

Flux v2 would need to be setup with a Kustomization CR (it can dynamically build it's own kustomization.yaml). The entire thing will fail if cert-manager is not installed but Issuers are in the manifests. For chicken and egg issues where CR's don't exist yet, I just built a shim that would wrap the problem inside [HelmReleases](https://github.com/fossas/charts-util/tree/master/charts/manifest-shim), which can depend on each other.

In fact, being forced to this state, I've found that I actually like the way my problems had to be addressed in the Flux v2 eco system. My Kustomization does gitops by using a repo where all objects are required to be a HelmRelease. Automation for pull requests validate HelmReleases with kubeconform and yq checking the "kind". PR's can't be merged  if validation fails.

Not saying Argo is bad by any stretch here, but having more field of view over this, it does fall a bit short in ways I didn't expect.

**Comment by Xosmond** [comment]
@protosam Thank you.
It's very weird that so many people have to deal with this issue without a proper response from the maintainer, clear error root cause or updated documentation. I'll look for an alternative tool as I don't plan to spent my time debugging a tool instead of my clusters/applications.

**Comment by JMarji** [comment]
So I was having trouble with creating an app in ArgoCD after successfully connecting to my repo over ssh in ArgoCD. 

I was using bitbucket as a repository. by reading the error message that ArgoCD output when I hit app create. Which went something like "argocd@bitbucket.org unable to "git fetch... etc" (Public Key)". I realized that Argo was trying to authenticate with bitbucket using the username argocd. This would not work in my setup as we did not have a user account set up for argocd. 

TLDR; I selected https instead of ssh in ArgoCD and used bitbucket app passwords in account settings on bitbucket  and used my bitbucket username and the app password provided through the app password process. solved all my auth problems with the repo.

**Comment by diptripa** [comment]
> ```
> ssh-keygen -t ed25519 -C "argocd" -f argocd
> ```
> 
> from #7600 Has allowed me to finally craft a key argocd will use, it's not the same host (e.g not github).
> 
> I think argocd isn't supporting the default flags of whatever ssh-keygen arch installs.

This really helped solve this problem! Thanks

**Comment by blakepettersson** [comment]
See [my comment](https://github.com/argoproj/argo-cd/issues/7600#issuecomment-1367788996) in #7600.

> This issue might be related to https://github.com/golang/go/issues/37278, specifically https://github.com/golang/go/issues/37278#issuecomment-969797827. https://github.com/golang/crypto/pull/211 might address this, but unknown when/if this will be merged...

**Comment by blakepettersson** [comment]
golang/crypto@6fad3dfc may have fixed this. Argo CD 2.7.x upgraded to `golang.org/x/crypto v0.6.0` so might be worth trying out there?

**Comment by urmanchev** [comment]
In my case I had to remove passphrase from a key and then it started working. I figured that out by installing argo cli and then adding repo manually specifying a key path and there was a clear error about passphrase protection.

`argocd repo add REPO_URL_HERE --ssh-private-key-path ~/.ssh/argocd`

> FATA[0000] rpc error: code = Unknown desc = error testing repository connectivity: ssh: this private key is passphrase protected

**Comment by horlyk** [comment]
In my case I had to remove passphrase from a key and then it started working. I figured that out by installing argo cli and then adding repo manually specifying a key path and there was a clear error about passphrase protection.

`argocd repo add REPO_URL_HERE --ssh-private-key-path ~/.ssh/argocd`

> FATA[0000] rpc error: code = Unknown desc = error testing repository connectivity: ssh: this private key is passphrase protected

**Comment by BacLuc** [comment]
In my case the URL had a space in front of the git url.
When i looked into the repo secret, this was the problem:

```
Name:         repo-18972354
Namespace:    argocd
Labels:       argocd.argoproj.io/secret-type=repository
Annotations:  managed-by: argocd.argoproj.io

Type:  Opaque

Data
====
url: " git@host:group/repo.git"
name: asdf
project: myproject
sshPrivateKey: -----BEGIN OPENSSH PRIVATE KEY-----
CENSORED
-----END OPENSSH PRIVATE KEY-----
type: git
```

**Comment by erwinkramer** [comment]
In my case, my private key was CRLF formatted, which succeeds initially when adding a repo, but fails when you try to use it. When I changed the format to LF, it all worked.
